rust/noosphere-storage/src/implementation/rocks_db.rs

```diff
@@ -0,0 +1,212 @@
+use crate::{storage::Storage, store::Store, SPHERE_DB_STORE_NAMES};
+use anyhow::{anyhow, Result};
+use async_trait::async_trait;
+use lazy_static::lazy_static;
+use rocksdb::{ColumnFamilyDescriptor, DBWithThreadMode, Options};
+use std::{
+    collections::HashMap,
+    path::{Path, PathBuf},
+    sync::{Arc, Mutex},
+};
+
+#[cfg(not(feature = "rocksdb-multi-thread"))]
+type DbInner = DBWithThreadMode<rocksdb::SingleThreaded>;
+#[cfg(not(feature = "rocksdb-multi-thread"))]
+type ColumnType<'a> = &'a rocksdb::ColumnFamily;
+#[cfg(feature = "rocksdb-multi-thread")]
+type DbInner = DBWithThreadMode<rocksdb::MultiThreaded>;
+#[cfg(feature = "rocksdb-multi-thread")]
+type ColumnType<'a> = Arc<rocksdb::BoundColumnFamily<'a>>;
+
+lazy_static! {
+    static ref DB_HANDLES: Mutex<HashMap<PathBuf, std::sync::Weak<DbInner>>> =
+        Mutex::new(HashMap::default());
+}
+
+/// A RocksDB implementation of [Storage].
+///
+/// Caveats:
+/// * Values are limited to 4GB(?) [https://github.com/facebook/rocksdb/wiki/Basic-Operations#reads]
+/// * Potential workarounds of cf_handles not being [Sync] [https://github.com/rust-rocksdb/rust-rocksdb/issues/407]
+#[derive(Clone, Debug)]
+pub struct RocksDbStorage {
+    db: Arc<DbInner>,
+    #[allow(unused)]
+    path: PathBuf,
+}
+
+impl RocksDbStorage {
+    pub fn new(path: PathBuf) -> Result<Self> {
+        std::fs::create_dir_all(&path)?;
```

Comment:
Might be safer to canonicalize first?

Reply:
Cannot canonicalize a path that does not exist